Implemented Secret Room "Forgotten Study" (Issue #102)

Sofiene-Bembli opened the pull request:
This pull request adds the new secret room **Forgotten Study** as described in Issue #102.

**-  What was implemented :**

- Created a new room class **" ForgottenStudy "** in **" zdd_rooms.py ".**
- Added a hidden entrance that becomes accessible after inspecting the archive.
- Added a simple puzzle involving the locked wooden box.
- Puzzle requires the **"old book"** item to solve.
        
**_- When solved:_**
    -  The old book is removed from the player inventory.
    - The box opens and rewards the player with the **"exam cheat sheet"** item.
               
- Added the room to the cellar floor in **" zdd_adventure.py ".**
- Updated the archive room to reveal the secret passage.

Sofiene-Bembli committed:
"Added Forgotten Study room (Issue #102)"

Emen-Fouda commented on adventure/zdd_rooms.py:
```diff
@@ -13,6 +13,61 @@ def run_story(self, user_items):
+
+            # Give cheat sheet
+            print("You obtained the 'exam cheat sheet'!")
+            user_items.append(CHEAT_SHEET)
```
The code uses a variable called `CHEAT_SHEET` when the puzzle is solved. I couldn't find where this variable is defined or imported, so I think this might cause the game to **crash.**

Is the `CHEAT_SHEET` item defined in your other Pull Request? If so, we might need to add an import statement to this file to make it work.

Emen-Fouda replied:
I noticed that when the puzzle is solved, the code tries to give the player the `CHEAT_SHEET` item. When I tested this, the game crashed with a `NameError` because this variable isn't defined in `zdd_rooms.py`. 

Is this item defined in your other Pull Request? If so, this file will need to import it to fix the **crash."**

Emen-Fouda commented on adventure/zdd_rooms.py:
```diff
@@ -13,6 +13,61 @@ def run_story(self, user_items):
+            print("You leave the box untouched.")
+            return user_items
+
+class ArchiveWithSecret(Room):
+    def show_items(self, user_items):
+        print("You inspect the archive… old folders, dust… and a LARGE BOOKSHELF?")
+        print("You pull it slightly… A secret door opens to the FORGOTTEN STUDY!")
+        return super().show_items(user_items)
```
I see you've used method overriding on `show_items` in the `ArchiveWithSecret` class to reveal the secret passage.

That's a clever way to do it, however after the message is printed, the player has no way to actually enter the `Forgotten Study`. The room is never added as a valid exit from the archive.

**Suggestion:** To fix this, the `show_items` method should probably also add the `Forgotten study` to the current floor's rooms dictionary. This would make the enter forgotten study command work right after the player discovers the secret."

Emen-Fouda reviewed (changes requested):
This is a great feature, and with these fixes, it will be a nice addition to the game.

Sofiene-Bembli committed:
Add 'study' room to the cellar

Sofiene-Bembli committed:
Merge branch 'main' into issue_102

Sofiene-Bembli commented:
@Emen-Fouda Thanks a lot for the detailed review!
I have now:

- defined the CHEAT_SHEET item in zdd_rooms.py, so the NameError is fixed

- added an instance of ForgottenStudy to ALL_ROOMS with the label "study"

- added cellar.add_room("study", ALL_ROOMS["study"]), so the player can enter the room after discovering the secret passage.

Feel free to review again, I'm happy to adjust further if something is unclear.

Sofiene-Bembli committed:
Small correction

Sofiene-Bembli committed:
small correction

Emen-Fouda reviewed (approved):
Great job.

Emen-Fouda commented:
@julianpollmann pull request ready to be merged
